codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -0,0 +1,686 @@
+          );
+      }
+      return """
+      return func () *%s {
+          if %s == nil {
+              return nil;
+          }
+          return &%s{%s.(*%s)}
+      }()""".formatted(
+          namespace.concat(
+            context.symbolProvider().toSymbol(serviceShape).getName()
+          ),
+          dataSource,
+          namespace.concat(
+            context.symbolProvider().toSymbol(serviceShape).getName()
+          ),
+          dataSource,
+          DafnyNameResolver.getDafnyClient(serviceShape.toShapeId().getName())
+        );
+    }
+  }
+
+  @Override
+  protected String getDefault(final Shape shape) {
+    throw new CodegenException(
+      String.format(
+        "Unsupported conversion of %s to %s using the %s protocol",
+        shape,
+        shape.getType(),
+        context.protocolGenerator().getName()
+      )
+    );
+  }
+
+  @Override
+  public String blobShape(final BlobShape shape) {
+    writer.addImportFromModule(DAFNY_RUNTIME_GO_LIBRARY_MODULE, "dafny");
+    return """
+    return func () []byte {
+    var b []byte
+    if %s == nil {
+        return nil
+    }
+    for i := dafny.Iterate(%s) ; ; {
+        val, ok := i()
+        if !ok {
+            return b
+        } else {
+            b = append(b, val.(byte))
+        }
+    }
+    }()""".formatted(dataSource, dataSource);
+  }
+
+  @Override
+  public String structureShape(final StructureShape shape) {
+    if (shape.hasTrait(ReferenceTrait.class)) {
+      return referenceStructureShape(shape);
+    }
+    final var builder = new StringBuilder();
+    writer.addImportFromModule(
+      SmithyNameResolver.getGoModuleNameForSmithyNamespace(
+        shape.toShapeId().getNamespace()
+      ),
+      DafnyNameResolver.dafnyTypesNamespace(shape)
+    );
+    final String referenceType = (this.isOptional) ? "&" : "";
+    builder.append(
+      "return %1$s%2$s{".formatted(
+          referenceType,
+          SmithyNameResolver
+            .smithyTypesNamespace(shape)
+            .concat(".")
+            .concat(shape.getId().getName())
+        )
+    );
+    for (final var memberShapeEntry : shape.getAllMembers().entrySet()) {
+      final var memberName = memberShapeEntry.getKey();
+      final var memberShape = memberShapeEntry.getValue();
+      final var targetShape = context
+        .model()
+        .expectShape(memberShape.getTarget());
+      //TODO: Is it ever possible for structure to be nil?
+      String maybeAssertion = "";
+      if (dataSource.equals("input")) {
+        maybeAssertion =
+          ".(".concat(
+              DafnyNameResolver.getDafnyType(
+                shape,
+                context.symbolProvider().toSymbol(shape)
+              )
+            )
+            .concat(")");
+      }
+      final boolean assertionRequired =
+        memberShape.isOptional() &&
+        targetShape.isStructureShape() &&
+        !targetShape.hasTrait(ReferenceTrait.class);
+      final var derivedDataSource =
```

Comment:
In general I'm struggling to understand what the generated code looks like in context, but this is probably the toughest one for me to grasp.

I think the easiest thing to do is link to some generated code? Do you have an MPL branch I can look at?

I don't know if this works with your timeline, but it would also help if you could leave comments above complicated code blocks with an inline example of what the generated code looks like, and maybe leave some more guiding comments?

Reply:
Yeah it's hard to get a feel for generated code when there's so many template params to be filled in. But adding generated code might pollute the codegen and also end up becoming stale.

Here's a link the MPL that might give you some context: https://github.com/aws/aws-cryptographic-material-providers-library/blob/Golang/dev/AwsCryptographyPrimitives/runtimes/go/ImplementationFromDafny-go/awscryptographyprimitivessmithygenerated/to_native.go#L67

Reply:
Thanks! This is helpful